modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java

```diff
@@ -327,9 +331,15 @@ protected void onNodeLeft(UUID evtNodeId) {
     protected void enqueue(Collection<?> col) {
         assert Thread.holdsLock(this);
 
-        queue.add((Collection<R>)col);
-
-        cnt.addAndGet(col.size());
+        if(limitCnt <= 0 || limitCnt >= col.size()) {
+            queue.add((Collection<R>)col);
+            cnt.addAndGet(col.size());
+            limitCnt -= col.size();
+        } else {
+            int toAdd = limitCnt;
+            queue.add(new ArrayList(col).subList(0, toAdd));
```

Comment:
I used `new ArrayList()` because array size here will always be smaller then `pageSize`. Usually it's around 1000 (please correct me if no). This is small array and there should not be performance impact. Working with streams seems to be more resource consuming in this case.
Please consider what should match here, I will update the code.

Reply:
Don't think that we will deal with huge arrays here ever. So why to add code complexity?)

Reply:
Unfortunately it is a common practice in Ignite. We do not have cheap performance measurement feedback loop, consequently we are trying to use faster approaches but possibly sacrificing code simplicity. There were a lot of cases from a real practice where a single "innocent" code line (not lock and not system call) caused visible performance drop.

Also I can thing about another option. Apparently a collection coming to `enqueue` method is always a `List`. We can try to do a refactoring and change an argument type to `List`. Then `List.sublist` method could be used directly. Possibly, it deserves a separate ticket. What do you think?